From 4eb6e4d959646d6d03949d12a6b4fa801b82b826 Mon Sep 17 00:00:00 2001
From: Igor Goltsov <igor@ontapgroup.com>
Date: Tue, 23 Nov 2021 15:01:25 +0300
Subject: [PATCH] MPGS-603: [Merchant][Github][Target][M2] Payment form is
 still rendered even when the payment session JS fails to load entirely.

---
 i18n/en_GB.csv                                |  1 +
 i18n/en_US.csv                                |  1 +
 view/frontend/web/css/source/_module.less     | 27 ++++++++++++
 .../postponed-adapter-activator-factory.js    | 41 ++++++++++++++++++
 .../view/payment/method-renderer/tns-hpf.js   | 42 +++++++++++++++----
 .../web/template/payment/hpf-cc-form.html     |  2 +-
 .../web/template/payment/tns-hpf.html         | 11 ++---
 7 files changed, 112 insertions(+), 13 deletions(-)
 create mode 100644 view/frontend/web/css/source/_module.less
 create mode 100644 view/frontend/web/js/lib/postponed-adapter-activator-factory.js

diff --git a/i18n/en_GB.csv b/i18n/en_GB.csv
index 89fbf2e..f0233f2 100644
--- a/i18n/en_GB.csv
+++ b/i18n/en_GB.csv
@@ -102,3 +102,4 @@ bankAccountHolder,"Account Holder"
 bankAccountNumber,"Account Number"
 routingNumber,"Routing Number"
 secCode,"SEC Code"
+It is impossible to continue with this Payment Method. Please try again later.,It is impossible to continue with this Payment Method. Please try again later.
diff --git a/i18n/en_US.csv b/i18n/en_US.csv
index 89fbf2e..f0233f2 100644
--- a/i18n/en_US.csv
+++ b/i18n/en_US.csv
@@ -102,3 +102,4 @@ bankAccountHolder,"Account Holder"
 bankAccountNumber,"Account Number"
 routingNumber,"Routing Number"
 secCode,"SEC Code"
+It is impossible to continue with this Payment Method. Please try again later.,It is impossible to continue with this Payment Method. Please try again later.
diff --git a/view/frontend/web/css/source/_module.less b/view/frontend/web/css/source/_module.less
new file mode 100644
index 0000000..9d07e3d
--- /dev/null
+++ b/view/frontend/web/css/source/_module.less
@@ -0,0 +1,27 @@
+/*
+ * Copyright (c) 2021 Mastercard
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+.checkout-payment-method {
+    .payment-method._active {
+        .payment-method-content {
+            .form {
+                &.hidden {
+                    display: none;
+                }
+            }
+        }
+    }
+}
diff --git a/view/frontend/web/js/lib/postponed-adapter-activator-factory.js b/view/frontend/web/js/lib/postponed-adapter-activator-factory.js
new file mode 100644
index 0000000..6d938c2
--- /dev/null
+++ b/view/frontend/web/js/lib/postponed-adapter-activator-factory.js
@@ -0,0 +1,41 @@
+/*
+ * Copyright (c) 2016-2021 Mastercard
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/*global define*/
+define(['jquery'], function ($) {
+    return function (activateAdapterCallback) {
+        var isFormRenderedPromise = $.Deferred();
+        var isAdapterLoadedPromise = $.Deferred();
+
+        $.when(
+            isFormRenderedPromise,
+            isAdapterLoadedPromise
+        ).then(activateAdapterCallback);
+
+        return {
+            setIsFormRendered: function () {
+                if (isFormRenderedPromise.state() !== 'resolved') {
+                    isFormRenderedPromise.resolve();
+                }
+            },
+
+            setIsAdapterLoaded: function () {
+                if (isAdapterLoadedPromise.state() !== 'resolved') {
+                    isAdapterLoadedPromise.resolve()
+                }
+            },
+        }
+    }
+});
diff --git a/view/frontend/web/js/view/payment/method-renderer/tns-hpf.js b/view/frontend/web/js/view/payment/method-renderer/tns-hpf.js
index 3ae88ce..86f5f5c 100644
--- a/view/frontend/web/js/view/payment/method-renderer/tns-hpf.js
+++ b/view/frontend/web/js/view/payment/method-renderer/tns-hpf.js
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2016-2020 Mastercard
+ * Copyright (c) 2016-2021 Mastercard
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -23,7 +23,8 @@ define(
         'Magento_Checkout/js/action/set-payment-information',
         'uiLayout',
         'Magento_Checkout/js/model/full-screen-loader',
-        'Magento_Vault/js/view/payment/vault-enabler'
+        'Magento_Vault/js/view/payment/vault-enabler',
+        'OnTap_MasterCard/js/lib/postponed-adapter-activator-factory',
     ],
     function (
         $,
@@ -33,7 +34,8 @@ define(
         setPaymentInformationAction,
         layout,
         fullScreenLoader,
-        VaultEnabler
+        VaultEnabler,
+        postponedAdapterActivatorFactory
     ) {
         'use strict';
 
@@ -42,6 +44,7 @@ define(
                 template: 'OnTap_MasterCard/payment/tns-hpf',
                 active: false,
                 adapterLoaded: false,
+                showCcForm: false,
                 buttonTitle: null,
                 buttonTitleEnabled: $t('Place Order'),
                 buttonTitleDisabled: $t('Please wait...'),
@@ -49,16 +52,18 @@ define(
                     onActiveChange: 'active'
                 },
                 creditCardExpYear: '',
-                creditCardExpMonth: ''
+                creditCardExpMonth: '',
             },
             placeOrderHandler: null,
             validateHandler: null,
+            adapterActivator: null,
             sessionId: null,
 
             initialize: function () {
                 this._super();
                 this.vaultEnabler = VaultEnabler();
                 this.vaultEnabler.setPaymentCode(this.getVaultCode());
+                this.adapterActivator = postponedAdapterActivatorFactory(this.activatePaymentAdapter.bind(this))
 
                 return this;
             },
@@ -82,7 +87,8 @@ define(
                         'adapterLoaded',
                         'creditCardExpYear',
                         'creditCardExpMonth',
-                        'buttonTitle'
+                        'buttonTitle',
+                        'showCcForm',
                     ]);
 
                 this.buttonTitle(this.buttonTitleDisabled);
@@ -171,14 +177,18 @@ define(
 
             loadAdapter: function () {
                 var config = this.getConfig();
-                require([config.component_url], this.paymentAdapterLoaded.bind(this));
+                require(
+                    [config.component_url],
+                    this.setIsAdapterLoaded.bind(this),
+                    this.paymentAdapterLoadFailed.bind(this)
+                );
             },
 
             isCheckoutDisabled: function () {
                 return !this.adapterLoaded() || !this.isPlaceOrderActionAllowed();
             },
 
-            paymentAdapterLoaded: function () {
+            activatePaymentAdapter: function () {
                 this.isPlaceOrderActionAllowed(false);
                 this.buttonTitle(this.buttonTitleDisabled);
 
@@ -187,6 +197,7 @@ define(
                     frameEmbeddingMitigation: ['x-frame-options'],
                     callbacks: {
                         initialized: function () {
+                            this.showCcForm(true);
                             this.adapterLoaded(true);
                             this.isPlaceOrderActionAllowed(true);
                         }.bind(this),
@@ -195,6 +206,15 @@ define(
                 }, this.getId());
             },
 
+            paymentAdapterLoadFailed: function() {
+                this.messageContainer.addErrorMessage({
+                    message: $t('It is impossible to continue with this Payment Method. Please try again later.')
+                });
+                this.buttonTitle(this.buttonTitleEnabled);
+                this.adapterLoaded(false);
+                this.isPlaceOrderActionAllowed(false);
+            },
+
             formSessionUpdate: function (response) {
                 var fields = this.getCardFields();
                 for (var field in fields.card) {
@@ -330,6 +350,14 @@ define(
 
             threeDSecureCancelled: function () {
                 this.isPlaceOrderActionAllowed(true);
+            },
+
+            setIsCcFormRendered: function () {
+                this.adapterActivator.setIsFormRendered();
+            },
+
+            setIsAdapterLoaded: function () {
+                this.adapterActivator.setIsAdapterLoaded();
             }
         });
     }
diff --git a/view/frontend/web/template/payment/hpf-cc-form.html b/view/frontend/web/template/payment/hpf-cc-form.html
index bc5453d..d8ef329 100644
--- a/view/frontend/web/template/payment/hpf-cc-form.html
+++ b/view/frontend/web/template/payment/hpf-cc-form.html
@@ -14,7 +14,7 @@
   ~ limitations under the License.
   -->
 
-<fieldset data-bind="attr: {class: 'fieldset payment items ccard ' + getCode(), id: 'payment_form_' + getCode()}">
+<fieldset data-bind="afterRender: setIsCcFormRendered, attr: {class: 'fieldset payment items ccard ' + getCode(), id: 'payment_form_' + getCode()}">
     <div class="field number required">
         <label data-bind="attr: {for: getCode() + '_cc_number'}" class="label">
             <span><!-- ko i18n: 'Credit Card Number'--><!-- /ko --></span>
diff --git a/view/frontend/web/template/payment/tns-hpf.html b/view/frontend/web/template/payment/tns-hpf.html
index 320ca6e..f7c8f00 100644
--- a/view/frontend/web/template/payment/tns-hpf.html
+++ b/view/frontend/web/template/payment/tns-hpf.html
@@ -1,5 +1,5 @@
 <!--
-  ~ Copyright (c) 2016-2019 Mastercard
+  ~ Copyright (c) 2016-2021 Mastercard
   ~
   ~ Licensed under the Apache License, Version 2.0 (the "License");
   ~ you may not use this file except in compliance with the License.
@@ -32,10 +32,11 @@
             <!--/ko-->
         </div>
 
-        <form class="form" id="co-transparent-form" action="#" method="post" data-bind="mageInit: {
-            'transparent':{
-                'context': context()
-            }, 'validation':[]}">
+        <form class="form" id="co-transparent-form" action="#" method="post"
+              data-bind="css: {'hidden': !showCcForm()}, mageInit: {
+                'transparent':{
+                    'context': context()
+                }, 'validation':[]}">
             <!-- ko template: 'OnTap_MasterCard/payment/hpf-cc-form' --><!-- /ko -->
             <!-- ko if: (isVaultEnabled())-->
             <div class="field choice">
-- 
2.30.2

